### ISSUE TITLE
dom-change equivalent for iron-list

### ISSUE THREAD

**Issue by m4b** [PolymerElements/iron-list#62]
There doesn't seem to be an equivalent `dom-change` event (a la `<template is="dom-repeat">`) that's fired when iron-list has finished updating the list (i.e., _after_ the items have changed, and the list is done rendering the first visible batch).

An event like `dom-change` would be nice.

The reason is I find myself sometimes needing to perform batch-like post-processing on the items after they're stamped/rendered.

Perhaps this is unrealistic or at odds with the intent of the component, but just bringing this up here to see if anyone else thinks similarly.

**Comment by blasten** [comment]
Can you give an example that will require you to do that processing?

**Comment by m4b** [comment]
https://github.com/m4b/silicon-instructions

So I need to call `highlight.js` to get the prettified code.  Specifically this observer for the `instructions` array:

````javascript
    _newInstructions: function(instructions) {
      // add the highlighting into the instructions
      instructions.forEach(function(instr, index) {
        var res = hljs.highlightAuto(instr.mnemonic + ' ' + instr.op_str);
        instr.html = res.value; // this right here seems hackish
        instr.id = '_' + index; // so we can lookup the `<silicon-instruction>` easily
        instr.language = res.language;
      }, this);
    },
````
 Then in `silicon-instruction.html` I check to see if there is an `html` element, meaning it was post-processed, and stuff that into the innerHTML.  It all seems kind of hackish to me, dunno.

**Comment by blasten** [comment]
hmm, what about creating an element and using data binding to provide the input to the custom element?

**Comment by blasten** [comment]
ping

**Comment by m4b** [comment]
oh hehe, did you want me to do something?

I'm not sure what the best approach is; whether to have a notification when it's done rendering the current viewport, or preprocessing the item list via an observer and inserting what you need there.  I ended up sticking with the last and doing some other icky stuff for highlighting.  The current implementation is:

````js
    _newInstructions: function(instructions) {
      // add the highlighting into the instructions
      instructions.forEach(function(instr, index) {
        var res = hljs.highlightAuto(instr.mnemonic + ' ' + instr.op_str);
        instr.html = res.value;
        instr.id = '_' + index; // so we can lookup the `<silicon-instruction>` easily
        instr.language = res.language;
        instr.selected = ''; // more css hacks
      }, this);
    },
````
because now I need to do proper CSS highlighting, which is a combination of having a selected property inserted into the instruction and then using `this.classFollows`, which results in highlighting code like this:

````js
    _select: function(i) {
      var siliconInstruction = this.$$('#_' + i);
      var currentSiliconInstruction = this.$$('.selected');
      var inst = this.instructions[i];
      var oldInst = this.selected;
      this.$.selector.select(inst);
      if (siliconInstruction === currentSiliconInstruction) {
        siliconInstruction = undefined;
      }
      if (currentSiliconInstruction){
        this.toggleClass('glow', false, currentSiliconInstruction);
      }
      // moar silly hacks
      inst.selected = this.glow ? 'selected glow' : 'selected';
      if (oldInst) {
        oldInst.selected = '';
      }
      // swaps selectedness
      this.classFollows('selected', siliconInstruction, currentSiliconInstruction);
      // need to remove glow in case we switched from glow/non-glow
      if (this.glow) this.classFollows('glow', siliconInstruction, currentSiliconInstruction);
    },
````

Which is kind of manual and crappy, but couldn't figure out any other way to get around the re-used state aspect.

So I actually checked, and in iOS for dealing with these issues, especially really fast scrolling, you make a call like:

````objective-c
                UITableViewCell *cell = (id)[tableView cellForRowAtIndexPath:indexPath];
                if (cell){ // if the async request isn't fast enough, need to check if we're still onscreen
                    cell.imageView.image = image;
````
So each cell (i.e., list item) has an index; but if the scroll happens before the next line executes, the cell is null.  This is sort of what I do with `if (currentSiliconInstruction)`.

So, the answer is, I'm not sure what's best; it seems like you're following the iOS approach where all state is stored with the item, and woe unto others if not.  This does seem like the most maintainable; I just didn't like iterating over the array every change to stick in the extra state I needed due to performance considerations, since in my uses cases the instruction-list is going to change _extremely_ rapidly, so every change will cause `O(n)` processing...

**Comment by blasten** [comment]
what about if you have something like this:

```html
<iron-list>
  <template>
    <highlighter instructions="[[item.instructions]]"></highlighter>
  </template>
</iron-list>
```

**Comment by m4b** [comment]
Not sure what that is supposed to do, I'd have to see what highlighter does.  The current behavior is visible [here](http://www.m4b.io/silicon-instructions/components/silicon-instructions/demo)

**Comment by blasten** [comment]
In my example `<highlighter>` is basically your `<silicon-instruction>`. I can see your point.  Can you highlight each line independently of each other, or do you have to keep states. e.g.  scopes, blocks, contexts, etc. ?

**Comment by m4b** [comment]
So about half of the reason for this feature request was removed once you implemented the selection api (:thumbsup: btw), because I had to roll my own selector.

The other half is *sort of* still there, namely I need to preprocess each instruction.  But calling the highlighter during every scroll/redisplay of new items would certainly be less performant than preprocessing once on new items, in almost every use case, so I'd feel fine closing this issue.
